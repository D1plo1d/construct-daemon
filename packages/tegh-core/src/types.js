/* config */
export Config, { MockConfig } from './config/types/Config'
export * as AxisTypeEnum from './config/types/AxisTypeEnum'

/* jobQueue */
export * as JobHistoryTypeEnum from './jobQueue/types/JobHistoryTypeEnum'

export JobHistoryEvent from './jobQueue/types/JobHistoryEvent'
export JobQueueGraphQL from './jobQueue/types/JobQueue.graphql.js'
export Job, { MockJob } from './jobQueue/types/Job'
export JobGraphQL from './jobQueue/types/Job.graphql.js'
export JobFile, { MockJobFile } from './jobQueue/types/JobFile'

/* printer */
export PrinterGraphQL from './printer/types/Printer.graphql.js'

/* spool */
export * as PriorityEnum from './spool/types/PriorityEnum'
export * as TaskStatusEnum from './spool/types/TaskStatusEnum'
export * as ComponentTypeEnum from './config/types/ComponentTypeEnum'

export Task, { MockTask } from './spool/types/Task'
export TaskGraphQL from './spool/types/Task.graphql.js'
