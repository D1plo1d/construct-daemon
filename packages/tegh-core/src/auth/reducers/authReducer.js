import { loop, Cmd } from 'redux-loop'
import { Map, Record } from 'immutable'
import { ec as EC } from 'elliptic'

import displayInviteInConsole from '../sideEffects/displayInviteInConsole'

import { SET_CONFIG } from '../../config/actions/setConfig'

import createInvite from '../actions/createInvite'
import { DAT_PEER_HANDSHAKE_RECEIVED } from '../actions/datPeerHandshakeReceived'
import { DAT_PEER_DATA_RECEIVED } from '../actions/datPeerDataReceived'

const ec = new EC('curve25519')

export const initialState = Record({
  hostKeys: null,
  usersAndInvitesByPublicKey: Map(),
  datSessionsByPeerID: Map(),
})

const authReducer = (state = initialState, action) => {
  switch (action.type) {
    case SET_CONFIG: {
      const { config } = action.payload
      const { users, invites } = config.auth

      const admins = users.filter(user => user.admin)
      const adminInvites = invites.filter(invite => invite.admin)
      let consoleInvites = invites.filter(invite => invite.displayInConsole)

      const sideEffects = []

      /*
       * If no invites exist and no admins exist then create an invite so the
       * initial user can connect
       */
      if (admins.size === 0 && adminInvites.size === 0) {
        const nextAction = createInvite({
          displayInConsole: true,
          admin: true,
        })
        consoleInvites = consoleInvites.push(nextAction.payload.invite)
        sideEffects.push(Cmd.action(nextAction))
      }

      /*
       * Display invites in the console
       */
      consoleInvites.forEach(invite => (
        sideEffects.push(Cmd.run(displayInviteInConsole, {
          args: [{ hostPublicKey: hostKeys.getPublic(), invite }],
        }))
      ))

      return loop(state, Cmd.list(sideEffects))
    }
    case DAT_PEER_HANDSHAKE_RECEIVED: {
      const hostEphemeralKeys = ec.genKeyPair()
      hostEphemeralKeys.getPublic()
    }
    case DAT_PEER_DATA_RECEIVED: {
    }
    default: {
      return state
    }
  }
}

export default authReducer
