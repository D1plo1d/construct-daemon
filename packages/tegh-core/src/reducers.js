export config from './config/reducers/configReducer'
export devices from './devices/reducers/devicesReducer'
export jobQueue from './jobQueue/reducers/jobQueueReducer'
export log from './log/reducers/logReducer'
export macros from './macros/reducers/macrosReducer'
export pluginManager from './pluginManager/reducers/pluginManagerReducer'
export spool from './spool/reducers/spoolReducer'
export status from './printer/reducers/statusReducer'
