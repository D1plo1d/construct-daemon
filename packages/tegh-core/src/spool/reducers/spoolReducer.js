import { loop, Cmd } from 'redux-loop'
import {
  Record, List, Map,
} from 'immutable'

import isIdle from '../selectors/isIdle'

import {
  EMERGENCY,
  NORMAL,
  PREEMPTIVE,
  priorityOrder,
} from '../types/PriorityEnum'

import {
  PRINTING,
  isSpooled,
} from '../types/TaskStatusEnum'

/* printer actions */
import { PRINTER_READY } from '../../printer/actions/printerReady'
import { ESTOP } from '../../printer/actions/estop'
import { DRIVER_ERROR } from '../../printer/actions/driverError'
/* job actions */
import { CANCEL_JOB } from '../../jobQueue/actions/cancelJob'
import { DELETE_JOB } from '../../jobQueue/actions/deleteJob'
/* task actions */
import { SPOOL_TASK } from '../actions/spoolTask'
import requestDespool, { REQUEST_DESPOOL } from '../actions/requestDespool'
import despoolTask from '../actions/despoolTask'
import { CANCEL_TASK } from '../actions/cancelTask'
import cancelAllTasks, { CANCEL_ALL_TASKS } from '../actions/cancelAllTasks'

export const initialState = Record({
  priorityQueues: Record({
    [EMERGENCY]: List(),
    [PREEMPTIVE]: List(),
    [NORMAL]: List(),
  })(),
  tasks: Map(),
  currentTaskID: null,
})()

const removeTaskReferences = (state) => {
  const nextTaskIDs = state.tasks.map(task => task.id)

  let nextState = state
    .update('priorityQueues', queues => queues.map(queue => (
      queue.filter(taskID => nextTaskIDs.includes(taskID))
    )))

  if (nextTaskIDs.includes(state.currentTaskID) === false) {
    nextState = nextState.set('currentTaskID', null)
  }

  return nextState
}

const spoolReducer = () => (state = initialState, action) => {
  switch (action.type) {
    /* Spool reset actions */
    case PRINTER_READY:
    case ESTOP:
    case DRIVER_ERROR:
    case CANCEL_ALL_TASKS: {
      return initialState
    }
    case CANCEL_JOB:
    case DELETE_JOB: {
      const { jobID } = action.payload

      const nextState = state.update('tasks', tasks => (
        tasks.filter(task => task.jobID !== jobID)
      ))

      return removeTaskReferences(nextState)
    }
    case CANCEL_TASK: {
      const { taskID } = action.payload

      const nextState = state.update('tasks', tasks => tasks.remove(taskID))

      return removeTaskReferences(nextState)
    }
    case SPOOL_TASK: {
      const { payload } = action
      const { id, priority } = payload.task

      let nextState = state

      /*
      * if the task is an emergency then cancel all other tasks in the queue
      */
      if (priority === EMERGENCY) {
        nextState = initialState
      }

      nextState = state.setIn(['tasks', id], payload.task)

      if (
        isIdle.resultFunc(state.tasks) === false
        && priority !== EMERGENCY
        && payload.task.internal !== true
      ) {
        throw new Error('Cannot spool non-emergency tasks when printing a job')
      }


      /* create the task */
      const createAction = createTask({ task: payload.task })
      nextState = taskMap.createOne(nextState, createAction)

      const taskQueue = ['priorityQueues', priority]
      const shouldDespool = nextState.currentTaskID == null

      nextState = nextState
        .updateIn(taskQueue, list => list.push(id))

      /*
       * despool the first line if nothing is spooled
       */
      if (shouldDespool) {
        return loop(nextState, Cmd.action(requestDespool()))
      }

      return nextState
    }
    case REQUEST_DESPOOL: {
      const { priorityQueues, currentTaskID } = state
      let nextState = state
      if (currentTaskID != null) {
        /*
         * despool the next line or finish the task via the taskReducer
         */
        nextState = taskMap.updateOne(nextState, action, currentTaskID)
      }
      const currentTask = nextState.tasks.get(currentTaskID)
      if (currentTask == null || currentTask.status !== PRINTING) {
        /*
         * if the current task is done then despool the next task if there is
         * anything to despool.
         */
        const priority = priorityOrder.find(priorityOption => (
          priorityQueues[priorityOption].size > 0
        ))

        if (priority == null) {
          nextState = nextState.set('currentTaskID', null)
        } else {
          const nextTaskID = state.priorityQueues[priority].first()
          /*
           * start the task via the taskReducer
           */
          nextState = nextState
            .mergeIn(['tasks', nextTaskID], {
              startedAt: new Date().toISOString(),
              status: PRINTING,
              currentLineNumber: 0,
            })
            .set('currentTaskID', nextTaskID)
            .updateIn(['priorityQueues', priority], list => list.shift())
        }
      }

      if (nextState.currentTaskID == null) {
        return nextState
      }

      const nextTask = nextState.tasks.get(nextState.currentTaskID)
      return loop(nextState, Cmd.action(despoolTask(nextTask)))
    }
    case REQUEST_DESPOOL: {
      if (state.currentLineNumber < state.data.size - 1) {
        /*
         * if the task has more lines to execute then increment the line number
         */
        return state.update('currentLineNumber', i => i + 1)
      }
      /* mark tasks as done after they are completed */
      return state.merge({
        // TODO: stoppedAt should eventually be changed to be sent after
        // the printer sends 'ok' or 'error' and should be based off
        // estimated print time
        stoppedAt: new Date().toISOString(),
        status: DONE,
        /* the data of each completed task is deleted to save space */
        data: null,
      })
    }
    default: {
      return state
    }
  }
}

export default spoolReducer
