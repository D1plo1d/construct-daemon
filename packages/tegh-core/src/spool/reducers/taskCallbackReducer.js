import { loop, Cmd } from 'redux-loop'
import { Record, List } from 'immutable'

// TODO: implement task errored in spoolReducer
import { PRINTER_READY } from '../../printer/actions/printerReady'
import { ESTOP } from '../../printer/actions/estop'
import { DRIVER_ERROR } from '../../printer/actions/driverError'

import { TASK_ERRORED } from '../actions/taskErrored'
import { DESPOOL_TASK } from '../actions/despoolTask'
import { DESPOOL_COMPLETED } from '../actions/despoolCompleted'

export const initialState = List()

const spoolReducer = (state = initialState, action) => {
  switch (action.type) {
    case PRINTER_READY:
    case ESTOP:
    case DRIVER_ERROR: {
      return initialState
    }
    case TASK_ERRORED: {
      const { task } = action.payload

      if (task.onError) {
        return loop(
          state,
          Cmd.run(task.onError, {
            args: [task],
          }),
        )
      }

      return state
    }
    case DESPOOL_COMPLETED: {
      const { task, isLastLineInTask } = action.payload

      const nextState = state.shift()

      if (isLastLineInTask && task.onComplete) {
        return loop(
          nextState,
          Cmd.run(task.onComplete, {
            args: [task],
          }),
        )
      }

      return initialState
    }
    default: {
      return state
    }
  }
}

export default spoolReducer
