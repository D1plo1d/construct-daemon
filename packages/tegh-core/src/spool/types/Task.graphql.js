import tql from 'typiql'
import snl from 'strip-newlines'
import {
  GraphQLObjectType,
} from 'graphql'
import { GraphQLDate } from 'graphql-scalars'

import PrinterGraphQL from '../../printer/types/Printer.graphql.js'

import getTaskPercentComplete from '../selectors/getTaskPercentComplete'

const TaskType = new GraphQLObjectType({
  name: 'Task',
  description: 'A spooled set of gcodes to be executed by the printer',
  fields: () => ({
    id: {
      type: tql`ID!`,
    },
    name: {
      type: tql`String!`,
    },
    currentLineNumber: {
      type: tql`Int`,
    },
    totalLineNumbers: {
      type: tql`Int!`,
      resolve: source => source.data.size,
    },
    percentComplete: {
      // TODO: PercentageScalarType
      // type: tql`${PercentageScalarType}!`,
      type: tql`Float!`,
      args: {
        digits: {
          type: tql`Int!`,
          description: snl`
            The number of digits to the right of the decimal place to round to.
            eg.
            \`digits: 0\` => 83
            \`digits: 1\` => 82.6
            \`digits: 2\` => 82.62
          `,
        },
      },
      resolve: (source, { digits }, { store }) => {
        const state = store.getState().spool
        return getTaskPercentComplete(state)({
          taskID: source.id,
          digits,
        })
      },
    },
    createdAt: {
      type: tql`${GraphQLDate}!`,
    },
    startedAt: {
      type: tql`${GraphQLDate}`,
    },
    stoppedAt: {
      type: tql`${GraphQLDate}`,
    },
    status: {
      type: tql`String!`,
    },
    printer: {
      type: tql`${PrinterGraphQL}!`,
      resolve(_source, args, { store }) {
        const state = store.getState()
        return state
      },
    },
  }),
})

export default TaskType
