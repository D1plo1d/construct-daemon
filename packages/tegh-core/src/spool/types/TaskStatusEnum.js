export const SPOOLED = 'SPOOLED'
export const PRINTING = 'PRINTING'

export type TaskStatusEnumT =
  | 'SPOOLED'
  | 'PRINTING'

const TaskStatusEnum = {
  SPOOLED,
  PRINTING,
}

export default TaskStatusEnum
