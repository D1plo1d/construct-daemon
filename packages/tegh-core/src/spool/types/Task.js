import type { RecordOf, List as ListT } from 'immutable'
import uuid from 'uuid'
import { Record, List } from 'immutable'
import type { PriorityEnumT } from './PriorityEnum'
import type { TaskStatusEnumT } from './TaskStatusEnum'


import normalizeGCodeLines from '../../util/normalizeGCodeLines'
import { priorityOrder, NORMAL } from './PriorityEnum'
import { SPOOLED } from './TaskStatusEnum'

export type TaskT = RecordOf<{
  id: string,
  jobID: ?string,
  jobFileID: ?string,

  priority: PriorityEnumT,
  internal: boolean,
  data: ListT<string>,
  name: ?string,

  /*
   * The line number that will be executed by the printer by a saga after the
   * reducers run.
   */
  currentLineNumber: ?number,

  createdAt: ?number,
  startedAt: ?number,
  stoppedAt: ?number,
  status: TaskStatusEnumT,
}>

const TaskRecord = Record({
  id: null,
  priority: null,
  internal: null,
  name: null,
  jobID: null,
  jobFileID: null,
  data: null,
  status: null,
  currentLineNumber: null,
  createdAt: null,
  startedAt: null,
  stoppedAt: null,
})

const Task = (attrs) => {
  const {
    priority,
    internal,
    name,
    data,
  } = attrs

  if (typeof name !== 'string') {
    throw new Error('name must be a string')
  }
  if (typeof internal !== 'boolean') {
    throw new Error('internal must be a boolean')
  }
  if (!priorityOrder.includes(priority)) {
    throw new Error('invalid priority')
  }
  if (!Array.isArray(data)) {
    throw new Error('data must be an array of strings')
  }
  return TaskRecord({
    id: uuid.v4(),
    createdAt: new Date().toISOString(),
    status: SPOOLED,
    ...attrs,
    data: List(normalizeGCodeLines(data)),
  })
}

export const MockTask = attrs => Task({
  name: 'test.ngc',
  internal: true,
  priority: NORMAL,
  data: ['g1 x10', 'g1 y20'],
  ...attrs,
})

export default Task
