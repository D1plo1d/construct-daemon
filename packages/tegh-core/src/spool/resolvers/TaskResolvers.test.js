import typeDefs from '@tegh/schema'

import deterministicTestSetup from '../../util/testing/deterministicTestSetup'
import snapshotTestResolvers from '../../util/testing/snapshotTestResolvers'

import { initialState as spoolState } from '../reducers/spoolReducer'

import Task from '../types/Task'
import { NORMAL } from '../types/PriorityEnum'

describe('TaskResolvers', () => {
  it('presents a consistent API', () => {
    deterministicTestSetup()
    const task = Task({
      name: 'test task',
      internal: false,
      priority: NORMAL,
      data: ['g1 x10', 'g1 y20'],
    })

    const state = {
      spool: spoolState.setIn(['tasks', task.id], task),
      config: {
        printer: {
          id: 'test_printer_id',
        },
      },
    }

    snapshotTestResolvers({
      typeDefs,
      typeName: 'Task',
      rootValue: task,
      contextValue: {
        store: {
          getState: () => state,
        },
      },
    })
  })
})
