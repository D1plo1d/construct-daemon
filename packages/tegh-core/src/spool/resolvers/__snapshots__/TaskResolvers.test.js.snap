// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskResolvers presents a consistent API 1`] = `
Object {
  "_testTask": Object {
    "createdAt": "2018-01-01T00:00:00.000Z",
    "currentLineNumber": null,
    "id": "MOCK_ID_0",
    "name": "test task",
    "percentComplete": 0,
    "printer": Object {
      "id": "test_printer_id",
    },
    "startedAt": null,
    "status": "SPOOLED",
    "stoppedAt": null,
    "totalLineNumbers": 2,
  },
}
`;
